File: perl/pe014.pl
#!/usr/bin/perl
# http://projecteuler.net/problem=14
# Finds the starting number below 1 million that produces the longest chain
# following the Collatz sequence.

use strict;
use warnings;

my $limit = 1000000;
my @collatz;
my $max_steps = my $max_n = -1;

for (my $n = 1; $n < $limit; $n++) {
    my $steps = 1;

    for (my $i = $n; $i > 1; ) {
        # $i & 1 is equivalent to $i % 2
        if ($i & 1 == 1) {
            $i = (3 * $i) + 1;
            $steps++;
        }
        else {
            # $i >> 1 is equivalent to $i / 2
            $i >>= 1;

            if (defined($collatz[$i])) {
                $steps += $collatz[$i];
                last;
            }

            $steps++;
        }
    }

    $collatz[$n] = $steps;

    if ($steps > $max_steps) {
        $max_steps = $steps;
        $max_n = $n;
    }
}

print $max_n . "\n";

File: ruby/pe014.rb
#!/usr/bin/env ruby
# http://projecteuler.net/problem=14
collatz, max_n, max_steps = {}, 0, 0
(1..1_000_000).each do |n|
  steps, x = 1, n
  while x > 1
    if x & 1 == 1
      x = (3 * x) + 1
      steps += 1
    else
      x >>= 1
      if collatz.has_key? x
        steps += collatz[x]
        break
      end
      steps += 1
    end
  end
  collatz[n] = steps
  max_n, max_steps = n, steps if steps > max_steps
end
puts max_n
